RIOT-OS/RIOT
# make: introduce common BUILD_DIR

### Contribution description

This introduces a new environment variable for a common directory
that holds all output of the build process, such as application or
package binaries. This would also allow to easily redirect output
to any other location, e.g. for out-of-source builds.

~~The 2nd commit adds a `build` directory with subdirs `build/bin` and `build/pkg`. In the (near) future we might add something like `build/tests` to store test results.~~

~~The 3rd commit adds the APPLICATION name as a subdirectory to the bin output folder, this is similar to the pkg subdirectory hierarchy and will also allow to build multiple application in one out-of-source build directory (in the future).~~

### Testing procedure

Everything should still build as usual. ~~However, as some directories changed (see above) it might break something~~


### Issues/PRs references

- needed by #10095 and #10147
